### PR TITLE
New DAG engine initial checkin

### PR DESCRIPTION
**Pull request by HappyRay**
Initial DAG engine itself

#1486

The code is not used yet.

See earlier prototype and change history at #1487

More unit tests and refactoring will follow.

Node state machine diagram

azkaban dag state machine diagram - page 1

### CODE REVIEW

File: azkaban-exec-server/src/main/java/azkaban/dag/DagService.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Thread safe and non blocking service for DAG processing.
+ *
+ * <p>Allow external inputs to be given to a dag or node to allow the dag to transition states
+ * . Since only one thread is used to progress the DAG, thread synchronization is avoided.
+ */
+@SuppressWarnings("FutureReturnValueIgnored")
```
Comment (reallocf) [inline azkaban-exec-server/src/main/java/azkaban/dag/DagService.java]: Can you either resolve this warning or leave a comment with why it deserves to be suppressed? (I believe this is still the best practice, right?)

Comment (HappyRay) [reply azkaban-exec-server/src/main/java/azkaban/dag/DagService.java]: I think this is self-explanatory. The return value is not used.

File: azkaban-exec-server/src/main/java/azkaban/dag/Status.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.dag;
+
+import com.google.common.collect.ImmutableSet;
+
+enum Status {
```
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/dag/Status.java]: should we have two separated status enum for dag and node each?

File: azkaban-exec-server/src/main/java/azkaban/dag/Status.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  READY, // ready to run
+  DISABLED, // disabled by users. Treat as the node has the status of success
+  BLOCKED, // temporarily blocked. Need to be unblocked by another external event
```
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/dag/Status.java]: any example of node in blocked status?

File: azkaban-exec-server/src/test/java/azkaban/dag/DagTest.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+ * Tests the dag state ( including its nodes' states) transitions.
+ *
+ * Focuses on how the dag state changes in response to one external request.
+ */
+public class DagTest {
+
+  private final Dag testFlow = new Dag("fa", mock(DagProcessor.class));
+
+  @Test
+  public void dag_finish_with_only_disabled_nodes() {
```
Comment (chengren311) [inline azkaban-exec-server/src/test/java/azkaban/dag/DagTest.java]: why is public test method name separated by _? should we follow a consistent standard and document in the contributor guide?

File: azkaban-exec-server/src/main/java/azkaban/dag/Dag.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+ * A DAG (Directed acyclic graph) consists of {@link Node}s.
+ */
+class Dag {
+
+  private final String name;
+  private final DagProcessor dagProcessor;
+  private final List<Node> nodes = new ArrayList<>();
+  private Status status = Status.READY;
+
+  Dag(final String name, final DagProcessor dagProcessor) {
```
[the remark this answers is not in the document]
Comment (HappyRay) [reply azkaban-exec-server/src/main/java/azkaban/dag/Dag.java]: I plan to refactor this part later: separate construction of DAG from running a DAG.

I am interested in what you think of the different API choices. We can chat more off line.

[the remark this answers is not in the document]
Comment (HappyRay) [reply azkaban-exec-server/src/main/java/azkaban/dag/Dag.java]: Good question.

File: azkaban-exec-server/src/main/java/azkaban/dag/Status.java (hunk 2 above)
Comment (HappyRay) [reply azkaban-exec-server/src/main/java/azkaban/dag/Status.java]: Good question.

I thought about it.

The main factor that made me decide to use the same enum is to make testing easier. The same type can be put in the StatusChangeRecorder. 

That said, I can use string in the StatusChangeRecorder. Let's discuss more if it is worth refactoring.

File: azkaban-exec-server/src/main/java/azkaban/dag/Status.java (hunk 3 above)
Comment (HappyRay) [reply azkaban-exec-server/src/main/java/azkaban/dag/Status.java]: Not yet. It's not implemented.

File: azkaban-exec-server/src/main/java/azkaban/dag/Dag.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (HappyRay) [reply azkaban-exec-server/src/main/java/azkaban/dag/Dag.java]: Discussed offline, we agree it is ideal to make the DAG as immutable as possible.

File: azkaban-exec-server/src/main/java/azkaban/dag/Status.java (hunk 2 above)
Comment (HappyRay) [reply azkaban-exec-server/src/main/java/azkaban/dag/Status.java]: I will try it.

File: azkaban-exec-server/src/test/java/azkaban/dag/DagTest.java (hunk 4 above)
Comment (HappyRay) [reply azkaban-exec-server/src/test/java/azkaban/dag/DagTest.java]: Discussed offline, Cheng is fine with this style.

File: azkaban-exec-server/src/test/java/azkaban/dag/TestDagProcessor.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+  TestDagProcessor(final CountDownLatch dagFinishedLatch,
+      final StatusChangeRecorder statusChangeRecorder) {
+    this.dagFinishedLatch = dagFinishedLatch;
+    this.statusChangeRecorder = statusChangeRecorder;
+  }
+
+  @Override
+  public void changeStatus(final Dag dag, final Status status) {
+    System.out.println(dag);
```
Comment (chengren311) [inline azkaban-exec-server/src/test/java/azkaban/dag/TestDagProcessor.java]: system out

Comment (HappyRay) [reply azkaban-exec-server/src/test/java/azkaban/dag/TestDagProcessor.java]: This is intentional. 
Any downside?

Comment (HappyRay) [reply azkaban-exec-server/src/test/java/azkaban/dag/TestDagProcessor.java]: Discussed with Cheng offline. 
He agreed that it is ok to use in test code.

**Pull request merged**